### PR TITLE
Nexus

### PR DESCRIPTION
**Pull request by bergundy**
## What was changed

**EDIT**: Merged #1473 and #1475 into this PR, it now included the entire Nexus implementation for the SDK.

Added the `temporalnexus` package and implemented the handler side for Nexus, including registering and dispatching Nexus Operations.

Tests only pass with server `main`, so this PR should not be merged until the server is released.
A future PR will complete the nexus work allowing invoking Nexus Operations from a workflow.

See the [proposal](https://github.com/temporalio/proposals/blob/6db78a083ec8fe397527a3f0d7146c77c8204e96/nexus/sdk-go.md) for more information.

Also now memoizing `worker.Start()` to return consistent errors to callers and avoid rerunning the function unnecessarily.

Merge Checklist:

- [ ] Release Server
- [ ] Depend on tagged Nexus SDK

### CODE REVIEW

File: go.mod
Comment (cretz) [inline go.mod:5]: Curious, is this just an artifact of your tooling or was this change required?

File: go.mod
Comment (cretz) [inline go.mod:12]: I know we're not merging this PR until server tagged, but would also recommend this Nexus project is tagged before any sdk-go release with this dependency

File: internal/common/metrics/constants.go
Comment (cretz) [inline internal/common/metrics/constants.go:46]: I think putting all `Nexus` constants together is a bit easier to read than interleaved with their activity equivalents

File: internal/internal_nexus_task_handler.go
Comment (cretz) [inline internal/internal_nexus_task_handler.go:159]: I assume (inbound) interceptors are coming later?

File: internal/internal_nexus_task_handler.go
Comment (cretz) [inline internal/internal_nexus_task_handler.go:269]: Hrmm, I might have expected server to handle timeout and send cancellation. If a handler chooses not to respect timeout, what happens? If it is also handled server side, I think it's best to not also do it here except maybe with some considerable leeway to ensure server's cancellation logic is the one always processed.

File: internal/internal_nexus_worker.go
Comment (cretz) [inline internal/internal_nexus_worker.go:40]: Can probably just pass the two fields in to build these requests instead of pre-building a request and cloning just for these two fields

File: internal/internal_nexus_task_handler.go
Comment (cretz) [inline internal/internal_nexus_task_handler.go:353]: This is not a `nopReadCloser`, this `emptyReaderNopCloser`. This is going to return an EOF immediately (which is maybe what you want, though can change empty byte slice to `nil`)

File: internal/internal_nexus_task_poller.go
Comment (cretz) [inline internal/internal_nexus_task_poller.go:130]: If this is the case, confirm it. I am a bit concerned other errors will get swallowed.

File: internal/internal_nexus_worker.go
Comment (cretz) [inline internal/internal_nexus_worker.go:74]: Unrelated to this PR specifically, but sure wouldn't mind if this moved up to the aggregate worker instead of in each

File: internal/internal_worker.go
Comment (cretz) [inline internal/internal_worker.go:997]: I think this whole method should go inside memoized start. No need to repeat stuff above for each call.

File: internal/nexus_operations.go
Comment (cretz) [inline internal/nexus_operations.go:20]: Would call it `GetNexusOperationContext` similar to how the `activity` package has `GetX` calls that accept the context. Also, I would suggest not returning a boolean and just letting the result be nil if not in an operation context. Also, I would suggest making a new struct with fields from what's in context each time to discourage mutation.

Alternatively, consider `GetNexusOperationTemporalClient`, `GetNexusOperationInfo` (with just `TaskQueue` in it for now), and `GetNexusOperationLogger` to more consistently match what `activity` package does.

File: temporalnexus/operation.go
Comment (cretz) [inline temporalnexus/operation.go:1]: I assume calling-from-workflow stuff is coming later? Will it be in the `workflow` package? Will that code have to reference this package? I suggest considering adding it in this PR even in stub form to get circular imports correct.

**Review by cretz — COMMENTED** [review]
Looks great. Mostly minimal stuff.

File: go.mod
Comment (bergundy) [reply go.mod:5]: I just ran `go get ...` and `go mod tidy` but it may have been required due to the nexus sdk using `slog`.

File: go.mod
Comment (bergundy) [reply go.mod:12]: Sure, I can add this to a checklist.

Comment (bergundy) [reply go.mod:12]: Done.

File: internal/internal_nexus_task_handler.go
Comment (bergundy) [reply internal/internal_nexus_task_handler.go:159]: Yeah, I didn't include those for now.
I had them in a local branch but decided to start slim.

I'll make sure its tracked.

File: internal/internal_nexus_task_handler.go
Comment (bergundy) [reply internal/internal_nexus_task_handler.go:269]: We don't have sticky execution so there's not a way for the server to send cancelation. The server propagates this from the client request and also has its own context deadline.

I think it's good to cancel work that we know can't complete in time and have the server propagate this timeout.
As for whether the context deadline in the SDK should be shorter/same/longer than the one tracked on the server, fair point, but maybe shorter here is better so the SDK doesn't get a false sense of completion and the metrics we emit can be more accurate.

File: internal/internal_nexus_task_handler.go
Comment (bergundy) [reply internal/internal_nexus_task_handler.go:353]: Sure, I can rename. It can't be nil because the Nexus SDK tries to consume it.

File: internal/internal_nexus_task_poller.go
Comment (bergundy) [reply internal/internal_nexus_task_poller.go:130]: I'll just return the error just in case, thanks!

File: internal/internal_worker.go
Comment (bergundy) [reply internal/internal_worker.go:997]: Works for me, this was something that @Quinn-With-Two-Ns requested, so just confirming he's also okay with that.

File: internal/nexus_operations.go
Comment (bergundy) [reply internal/nexus_operations.go:20]: This is all internal so I wouldn't worry about it too much at the moment, we can expose some of this later as we see fit.

File: temporalnexus/operation.go
Comment (bergundy) [reply temporalnexus/operation.go:1]: Yeah, it'll be in the `workflow` package.
I started working on it and so far there aren't any circular deps.

File: internal/internal_nexus_worker.go
Comment (bergundy) [reply internal/internal_nexus_worker.go:40]: Yeah, I'll do that.

File: internal/common/metrics/constants.go
Comment (bergundy) [reply internal/common/metrics/constants.go:46]: Should I group the activity and workflow ones too?

Comment (cretz) [reply internal/common/metrics/constants.go:46]: Nah, I'd just add your stuff separate

File: internal/internal_nexus_task_handler.go
Comment (cretz) [reply internal/internal_nexus_task_handler.go:268]: But what happens in racy cases where SDK and server hit at the same time? If server-side happens first does that look the exact same as if the client-side one hit first and reported this failure back? It's important to have one system be the arbiter of true timeout errors. If you want a just-in-case for the other system, no prob, but having two separate systems that race each other to report timeout failure can result in racy inconsistencies.

File: internal/internal_nexus_task_handler.go
Comment (cretz) [reply internal/internal_nexus_task_handler.go]: I mean you can use `nil` instead of `[]byte{}`

File: internal/nexus_operations.go
Comment (cretz) [reply internal/nexus_operations.go:20]: Oh, my mistake. I think we will want to provide it to users one day, but if not exposed today, no prob

File: internal/internal_worker.go
Comment (Quinn-With-Two-Ns) [reply internal/internal_worker.go]: Users may retry starting their worker if it failed that is why I requested we don't memorize it, I think the likely hood is low, but it's very little effort to not memorize it so why not just avoid the breaking change?

Comment (bergundy) [reply internal/internal_worker.go]: I changed it to what @cretz suggested, I'm open to changing back.
I don't have a strong opinion here but slightly prefer memoizing the entire thing because it's easier to reason about.

File: internal/internal_nexus_task_handler.go
Comment (bergundy) [reply internal/internal_nexus_task_handler.go:268]: Keep in mind that clients also set deadlines on the request context.
I think giving the most up-to-date and accurate timeout to all of the processes involved in handling this request is preferable. That's how gRPC does it and this is essentially the SDK handling RPCs.

Also note that on context deadline errors we don't respond to the server, we just drop the task so I'm not as concerned with the racy inconsistency you're talking about.

**Review by Quinn-With-Two-Ns — APPROVED** [review]

**Comment by cretz** [comment]
Probably obvious, but let's not merge this until there's a server that works with it

File: go.mod
Comment (Quinn-With-Two-Ns) [reply go.mod:5]: Just FYI we should remove this before merging to main, it seems to mess up out CI trying to test multiple versions of Go

**Comment by bergundy** [comment]
Rebased and merged into the `nexus` branch.
I'll issue a separate PR to merge `nexus` into `main` once a server supporting Nexus is released.

**Pull request merged**
